app/login.go

```diff
@@ -55,6 +58,26 @@ func (a *App) AuthenticateUserForLogin(id, loginId, password, mfaToken string, l
 		return nil, err
 	}
 
+	// CWS login allow to use the one-time token to login the users when they're redirected to their
+	// installation for the first time
+	if IsCWSLogin(a, cwsToken) {
+		if err = checkUserNotBot(user); err != nil {
+			return nil, err
+		}
+		sessions, errSessions := a.GetSessions(user.Id)
+		if errSessions != nil {
+			return nil, errSessions
+		}
+		if len(sessions) > 0 {
```

Comment:
What's the purpose of this check? To make sure that this is the first login?  If so, woudn't the CWS login get re-enabled if a user simply logs out of all of their sessions? That's probably not what we want.

Reply:
Yeah that’s the idea. We have two ways of check that is the food login: sessions or status ‘LastActivityAt’ property. The status logic is optional using a flag so I ended up using sessions.

You’re right that if the user logs out you can enter again this way but that is a very small opportunity window. Yesterday talking with @gabrieljackson I was thinking on adding a another env bar with the expiration for the token, something like ‘CWS_CLOUD_TOKEN_EXPIRATION’ si after that time the token is not anymore. What do you think? // @jwilander

Reply:
The `loginId` (and thus `user`) comes from CWS, right? So an attacker who gains access to the CWS token could use it to login to any account that doesn't have currently active sessions? That's not good.

Reply:
Yeah but that’s the same problem that you have if someone gains access to the personal token for example right?

Reply:
Yeah, all other tokens are tied to a specific account, whereas this one gives universal access.

Reply:
Do think using an env var for the login_id should be enough? If not then we should move to use mmctl, modify it so we can pass a generated token, because we need to be have the token and mmctl doesn’t return the he enraged token,  limit  that to one account and then remove the token once is used

Reply:
I don't have a preference on the mechanism, be it env vars or mmctl, but I'd really like to see the token implementation be truly one-time instead of something close to it but not quite. And if the token is truly one-time then we don't need to care about the login ID either.

Reply:
The only two ways I can think of are:

- generate a token and insurer using mmctl, then remove/expire it after the use
- have three env. vars (token, login_id, expiration) in the installation

Do you see any problem in any of them? If no then we’re going to decide which option to use

Reply:
I'd be happier with explicit removal rather than time-based expiration. With a small enough expiration window I won't say no to the time-based option either, but there's a risk there.

Reply:
Ok then thank you very much for your help Juho. Then we should talk @jwilander, @gabrieljackson and me about this :)

Reply:
@ethervoid @jupenur could we do something like once the token is used here, store it in the MM database (the general purpose Tokens store would work) as "used" and don't allow login again with a "used" token? That should effectively make it one-time use

I'd prefer that instead of having mmctl generate it or making the CWS remove the env var since that introduces an extra hop which will be more complexity and more likely to fail

Reply:
Yes, that should do the trick. Didn't suggest that given that is kind of a hidden token flow and different of what we are using with the other tokens but I was working on something similar so all that I've done work for that solution too.

I'll suggest to use 64 characters token to match the other type of tokens

Reply:
Changes done and tested. I've added some tests too